### PR TITLE
DS-3489 Search REST endpoint Angular alignment

### PR DESCRIPTION
**Pull request by tomdesair**
This PR implements the changes defined in this REST contract PR https://github.com/DSpace/Rest7Contract/pull/21 to make the integration with the Angular UI easier.

To do this, we also had to address the issue of the inconsistent `_embedded` object structure.

### CODE REVIEW

File: dspace-api/src/main/java/org/dspace/discovery/SolrServiceImpl.java [hunk 1]
```diff
@@ -2216,4 +2258,29 @@ public FacetYearRange getFacetYearRange(Context context, DSpaceObject scope, Dis
         result.calculateRange(context, filterQueries, scope, this);
         return result;
     }
+
+    @Override
+    public String calculateExtremeValue(Context context, String valueField,
```
Comment (abollini) [inline dspace-api/src/main/java/org/dspace/discovery/SolrServiceImpl.java]: Is the SolrStats component a better choice here?

Comment (tomdesair) [reply dspace-api/src/main/java/org/dspace/discovery/SolrServiceImpl.java]: I'm not sure which SolrStats component you are referring to. Can you share the full class name?

Comment (abollini) [reply dspace-api/src/main/java/org/dspace/discovery/SolrServiceImpl.java]: I'm talking about the Stats Component. In our SOLR version tt can be used over numeric & date fields and provide direct access to the min and max values
http://localhost:8080/solr/search/select?q=*%3A*&rows=0&wt=json&indent=true&stats=true&stats.field=dc.date.accessioned_dt

File: dspace-api/src/main/java/org/dspace/discovery/configuration/DiscoverySearchFilterFacet.java [hunk 2]
```diff
@@ -55,4 +55,12 @@ public void setSortOrderSidebar(DiscoveryConfigurationParameters.SORT sortOrderS
     public String getFilterType() {
         return FILTER_TYPE_FACET;
     }
+
+    public boolean isExposeMinMax() {
```
Comment (tdonohue) [inline dspace-api/src/main/java/org/dspace/discovery/configuration/DiscoverySearchFilterFacet.java]: Could we add Javadocs here? I'm not sure I understand what "MinMax"is referring to.

File: dspace-spring-rest/src/main/java/org/dspace/app/rest/model/SearchConfigurationRest.java [hunk 3]
```diff
@@ -113,6 +116,28 @@ public int hashCode() {
         public static final String OPERATOR_CONTAINS = "contains";
         public static final String OPERATOR_NOTCONTAINS = "notcontains";
 
+        public boolean isHasFacets() {
```
Comment (tdonohue) [inline dspace-spring-rest/src/main/java/org/dspace/app/rest/model/SearchConfigurationRest.java]: Public methods need javadocs.

File: dspace-spring-rest/src/main/java/org/dspace/app/rest/model/SearchFacetEntryRest.java [hunk 4]
```diff
@@ -87,4 +96,28 @@ public int getFacetLimit() {
     public void setFacetLimit(final int facetLimit) {
         this.facetLimit = facetLimit;
     }
+
+    public boolean isExposeMinMax() {
```
Comment (tdonohue) [inline dspace-spring-rest/src/main/java/org/dspace/app/rest/model/SearchFacetEntryRest.java]: Again, Javadocs please

**Review by tdonohue — COMMENTED** [review]
👍 Code looks good overall, but similar to #2109, it could use some more Javadocs in some areas.  We should be striving to comment code well, minimally with Javadocs for public classes/methods.

**Review by terrywbrady — APPROVED** [review]
The code looks fine to me.  I tested locally, and I observed the presence of the minValue and maxValue.  I

**Comment by abollini** [comment]
@tomdesair please let us know if you like to revise the MinMax implementation using the Solr Stats component as by my suggestion
https://github.com/DSpace/DSpace/pull/2041#discussion_r201491773
or if you prefer to have this PR merged as soon as the @tdonohue concerns are solved and open an improvement issue for the MinMax implementation.

**Comment by tomdesair** [comment]
Yes @abollini I think it's better to move your alternative MinMax implementation to a new ticket as that would require more development work in the `dspace-api` module. 

We're working on documenting (dspace-api) methods where documentation is an added value.

**Comment by Raf-atmire** [comment]
@tdonohue I've written extra JavaDocs for public methods that have been added in this PR. If you find some time, could you have a look and see if they're okay or if anything slipped by, please?

**Review by tdonohue — APPROVED** [review]
👍 Looks good to me.  Thanks @Raf-atmire for the additional Javadocs.

@Raf-atmire or @tomdesair : I've just realized that I accidentally merged #2109 before this one (and it seems the dependency was the other way around).  My apologies.  

Should we rebase this and merge? Or does the #2109 merger include all the code here? (in which case we can likely just close this PR entirely)

**Review by tdonohue — CHANGES_REQUESTED** [review]
And I've just realized this PR has Travis failures.  The code looks good, but tests need fixing.

**Comment by Raf-atmire** [comment]
I'll fix this test and rebase the current master on this branch to ensure everything is merged properly. In order to preserve the history of comments here, I'll make a new PR and branch and mention it here shortly.

**Comment by tdonohue** [comment]
Closing, as most of the code in this PR was merged as part of #2109 (which was based off the same branch).  The code that wasn't already merged is now in #2163 (which is the followup to this PR).

**Pull request closed without merging**